src/porepy/numerics/fv/tpfa.py

```diff
@@ -266,3 +270,462 @@ def discretize(self, sd: pp.Grid, data: dict) -> None:
+
+    This is really a container class for methods that are needed to construct the
+    discretization. The actual discretization is considered a constitutive law, and can
+    be found in that part of the code.
+
+    """
+
+    def boundary_filters(
+        self,
+        mdg: pp.MixedDimensionalGrid,
+        subdomains: Sequence[pp.Grid],
+        boundary_grids: Sequence[pp.BoundaryGrid],
+        name: str,
+    ) -> tuple[pp.ad.Operator, pp.ad.Operator]:
+        """Filters for Dirichlet and Neumann boundary conditions.
+
+        Parameters:
+            mdg: Mixed-dimensional grid.
+            subdomains: List of grids.
+            boundary_grids: List of boundary grids.
+            name: Name of the operator.
+
+        Returns:
+            Tuple of Dirichlet and Neumann boundary filters.
+
+        """
+        dir_filter = pp.ad.TimeDependentDenseArray(
+            name=(name + "_filter_dir"), domains=boundary_grids
+        )
+        neu_filter = pp.ad.TimeDependentDenseArray(
+            name=(name + "_filter_neu"), domains=boundary_grids
+        )
+        proj = pp.ad.BoundaryProjection(mdg, subdomains, dim=1).boundary_to_subdomain
+        return proj @ dir_filter, proj @ neu_filter
+
+    def internal_boundary_filter(self, subdomains: list[pp.Grid]) -> pp.ad.DenseArray:
+        """Construct a dense array that filters out everything but internal boundaries.
+
+        Parameters:
+            subdomains: List of grids.
+
+        Returns:
+            DenseArray representation of the internal boundary filter.
+
+        """
+        is_internal = []
+        for sd in subdomains:
+            is_internal.append(sd.tags["fracture_faces"])
+        if len(is_internal) == 0:
+            fracture_faces = np.array([], dtype=int)
+        else:
+            fracture_faces = np.hstack(is_internal)
+
+        return pp.wrap_as_dense_ad_array(
+            fracture_faces, name="internal_boundary_filter"
+        )
+
+    def _block_diagonal_grid_property_matrix(
+        self,
+        domains: list[pp.Grid],
+        grid_property_getter: Callable[[pp.Grid], Any],
+    ) -> sps.spmatrix:
+        """Construct mapping matrix for the connectivity between two grids entities.
+
+        The mapping matrix is a block diagonal matrix where each block contains 1 where
+        the two entities are connected, and 0 otherwise.
+
+        Parameters:
+            domains: List of grids.
+            grid_property_getter: Function that returns the property of the grid that
+                should be used for the mapping.
+
+        Returns:
+            Mapping matrix.
+        """
+        if len(domains) == 0:
+            return sps.csr_matrix((0, 0))
+
+        blocks = []
+        for g in domains:
+            if g.dim == 0:
+                # 0d subdomains have no faces, so the projection might not exist.
+                # TODO: Implement this with proper handling of special cases (first and
```

Comment:
1. Don't we have to fix this?
2. Does this reveal we've never tested for intersections?

Reply:
1. Probably yes.
2. Correct.
Let me think a bit, but we may have to add one more test.

Reply:
See comments in a different channel.